src/msg/producer/writer/consumer_writer.go

```diff
@@ -448,12 +449,31 @@ func (w *consumerWriterImpl) reset(opts resetOptions) {
 }
 
 func (w *consumerWriterImpl) connectNoRetry(addr string) (io.ReadWriteCloser, error) {
-	conn, err := net.DialTimeout("tcp", addr, w.connOpts.DialTimeout())
+	// N.B.: this is roughly equivalent to what net.DialTimeout does; shouldn't introduce performance regressions.
+	ctx, cancel := context.WithTimeout(context.Background(), w.connOpts.DialTimeout())
+	defer cancel()
+
+	var (
+		conn net.Conn
+		err  error
+	)
+	if dialer := w.connOpts.ContextDialer(); dialer != nil {
+		conn, err = dialer(ctx, "tcp", addr)
```

Comment:
nit: if w.connOpts.ContextDialer() is not nil, just handle all the logic using the custom dialer here and return? instead of doing one of checks like L471 later, feels more readable that way

Reply:
Ah yeah I see what you're saying; my thinking for leaving it separated was that if the custom dialer *does* return a TCP connection, we might as well set up the TCP parameters in that case (since we can). I can switch it though if that behavior seems overly complicated.

Reply:
I prefer to keep them separated cleanly instead of having to zig-zag through all the if-else cases

Reply:
Oof now I'm looking at the code there's a critical bug here... (I returned instead of setting conn, err). Fixing that and adding more tests.

Reply:
Ok refactored a bit while keeping fundamentally same behavior as I had before. Not wedded to this; it could still probably be cleaner. One reason I like keeping the "if isTCPConn, set TCP settings" is that it can handle cases where dialer is simply `net.Dialer.DialContext`; saves a potential foot gun there

Potentially important change: I made the function always return `readWriterWithTimeout`, which it didn't before in the case where `keepAlive` wasn't set. Not sure if that was intended behavior before or a bug (seems like you'd still want timeout even without keep alive, but I could be wrong).